lib/Promises/Deferred.pm

```diff
@@ -4,186 +4,207 @@ package Promises::Deferred;
 
 use strict;
 use warnings;
-
-use Scalar::Util qw[ blessed reftype ];
-use Carp qw[ confess ];
-
+use Ref::Util qw/is_blessed_ref is_coderef/;
+use Promises::Deferred::Default;
 use Promises::Promise;
 
-use constant IN_PROGRESS => 'in progress';
-use constant RESOLVED    => 'resolved';
-use constant REJECTED    => 'rejected';
+use constant {
+    IN_PROGRESS => 'in progress',
+    RESOLVED    => 'resolved',
+    REJECTED    => 'rejected',
+};
+my %statuses= (
+    0 => IN_PROGRESS,
+    1 => RESOLVED,
+    2 => REJECTED
+);
+
+my @pending_callbacks;
+my $notify_sub= Promises::Deferred::Default->get_notify_sub;
+sub _set_backend {
+    my ( $class, $arg ) = @_;
+    my $backend = $arg->[0] or return;
+    unless ( $backend =~ s/^\+// ) {
+        $backend = 'Promises::Deferred::' . $backend;
+    }
 
-sub new {
-    my $class = shift;
-    bless {
-        resolved => [],
-        rejected => [],
-        status   => IN_PROGRESS
-    } => $class;
+    eval "use $backend; 1;" or return;
+    $notify_sub= $backend->get_notify_sub;
 }
 
-sub promise { Promises::Promise->new(shift) }
-sub status  { (shift)->{'status'} }
-sub result  { (shift)->{'result'} }
-
-# predicates for all the status possibilities
-sub is_in_progress { (shift)->{'status'} eq IN_PROGRESS }
-sub is_resolved    { (shift)->{'status'} eq RESOLVED }
-sub is_rejected    { (shift)->{'status'} eq REJECTED }
-sub is_done        { ! $_[0]->is_in_progress }
-
-# the three possible states according to the spec ...
-sub is_unfulfilled { (shift)->is_in_progress }
-sub is_fulfilled   { $_[0]->is_resolved }
-sub is_failed      { $_[0]->is_rejected }
+# The callback our backends will call; this will handle all the promises that have pending notifications
+sub _invoke_cbs_callback {
+    while (my $cb= shift @pending_callbacks) {
+        _invoke_cb($cb);
 
-sub resolve {
-    my $self = shift;
-
-    die "Cannot resolve. Already  " . $self->status
-        unless $self->is_in_progress;
-
-    $self->{'result'} = [@_];
-    $self->{'status'} = RESOLVED;
-    $self->_notify;
-    $self;
-}
-
-sub reject {
-    my $self = shift;
-    die "Cannot reject. Already  " . $self->status
-        unless $self->is_in_progress;
-
-    $self->{'result'} = [@_];
-    $self->{'status'} = REJECTED;
-    $self->_notify;
-    $self;
+        # Explicit undef. Why? Because this is where we're going to spend almost all of our time:
+        # deallocating Perl objects. So make it explicit as profilers will point people here.
+        undef $cb;
+    }
 }
 
-sub then {
-    my $self = shift;
-    my ( $callback, $error ) = $self->_callable_or_undef(@_);
-
-    my $d = ( ref $self )->new;
-    push @{ $self->{'resolved'} } => $self->_wrap( $d, $callback, 'resolve' );
-    push @{ $self->{'rejected'} } => $self->_wrap( $d, $error,    'reject' );
-
-    $self->_notify unless $self->is_in_progress;
-    $d->promise;
+sub new {
+    return bless \{cb=>[], state=>0}, __PACKAGE__;
```

Comment:
blessing into __PACKAGE__ is wrong and breaks subclassing, why are you not using $class (aka $_[0])?